### ISSUE TITLE
GraphicFill loading state when writing style

### ISSUE THREAD

**Issue by KaiVolland** [geostyler/geostyler-openlayers-parser#434]
As stated by @nevaldas the loading state of images can be a problem when writing GraphicFill styleFunctions with external graphics: https://github.com/geostyler/geostyler-openlayers-parser/pull/388#issuecomment-93328645 

Possible solutions could be:

- Wait for the images to be loaded before returning the style. This could be difficult when using placeholders in the image url.
- optionaly passing the olmap and add an onload listener to rerender it
- … 

Please discuss and add more ideas. :)

**Comment by jansule** [comment]
Also we could provide some kind of event, so the actual application can decide when to trigger the rerender. However, all applications would have to implement this then.

**Comment by geographika** [comment]
I ran into this in #649. I have a hacky workaround where after parsing the SLD file, it loops through and loads each of the images:

```js

function preloadIcon(src) {
    return new Promise((resolve) => {
        const icon = new Icon({ src, crossOrigin: 'anonymous' });
        icon.load();

        const interval = setInterval(() => {
            const state = icon.getImageState();
            if (state === OlImageState.LOADED || state === OlImageState.ERROR) {
                clearInterval(interval);
                resolve();
            }
        }, 20);

    });
}

async function preloadGraphicFillImages(gsStyle) {
    const urls = new Set();
    gsStyle.rules?.forEach((rule) => {
        rule.symbolizers?.forEach((s) => {
            if (s.kind === 'Fill' && s.graphicFill?.kind === 'Icon' && s.graphicFill.image) {
                urls.add(s.graphicFill.image);
            }
        });
    });
    await Promise.all([...urls].map(preloadIcon));
}

// in a processSldStyle function
...
    sldParser.readStyle(sldXmlString).then((gs) => {
        preloadGraphicFillImages(gs.output).then(() => {
...
```

I'm not sure how else to handle this without making all functions `async`. The above code is added at the application-level, and is specific to loading SLD so maybe moving it to the geostyler-sld-parser would be appropriate?

**Comment by KaiVolland** [comment]
This relies on caching mechanisms of the environment right? It seems like the most pragmatic solution, but from an architectural perspective i don't realy like it. 🤔 

Probably making the ol parsers methods async is the more sustainable approach. What do you think?

**Comment by geographika** [comment]
> This relies on caching mechanisms of the environment right? It seems like the most pragmatic solution, but from an architectural perspective i don't realy like it. 🤔
> 
> Probably making the ol parsers methods async is the more sustainable approach. What do you think?

Yes, making everything async would be nicer, but as I understand it, all the caller functions would need to be changed to async too. Something to maybe tackled in a codesprint.. 😄

**Comment by KaiVolland** [comment]
I added this issue to the todos: https://github.com/orgs/geostyler/discussions/2722
